windows-agent/internal/daemon/networking_windows.go

```diff
@@ -1,127 +1,78 @@
-	defer decorate.OnError(&err, "could not get network adapter addresses")
-
-	// This function is a wrapper around the Windows API GetAdaptersAddresses.
-	// https://learn.microsoft.com/en-us/windows/win32/api/iphlpapi/nf-iphlpapi-getadaptersaddresses
-	//
-	// This function takes in a buffer and fills it with a linked list.
-
-	// Flags from the Windows API.
-	//nolint:revive // Windows API constants are in shout case.
-	const (
-		GAA_FLAG_SKIP_ANYCAST       uint32 = 0x0002
-		GAA_FLAG_SKIP_MULTICAST     uint32 = 0x0004
-		GAA_FLAG_SKIP_DNS_SERVER    uint32 = 0x0008
-		GAA_FLAG_SKIP_FRIENDLY_NAME uint32 = 0x0010
-	)
-
-	// Return only IPv4 unicast addresses
-	const (
-		family uint32 = windows.AF_INET
-		flags  uint32 = GAA_FLAG_SKIP_ANYCAST | GAA_FLAG_SKIP_MULTICAST | GAA_FLAG_SKIP_DNS_SERVER | GAA_FLAG_SKIP_FRIENDLY_NAME
-	)
-
-	// We need a typed buffer rather than []byte because we don't want the GC to move
-	// the buffer around while we're using it, invalidating the NEXT pointers.
-	var buff buffer[windows.IpAdapterAddresses]
-
-	// Win32 API docs recommend a buff size of 15KB.
-	buff.resizeBytes(15 * kilobyte)
-
-	for range 10 {
-		size := buff.byteCount()
-		err := windows.GetAdaptersAddresses(family, flags, 0, &buff.data[0], &size)
-		if errors.Is(err, windows.ERROR_BUFFER_OVERFLOW) {
-			// Buffer too small, try again with the returned size.
-			buff.resizeBytes(size)
-			continue
-		}
-		if err != nil {
-			return nil, err
-		}
-
-		// The buffer is filled with the linked list of adapters, with the first element being the head.
-		// We return a pointer to the start of the buffer.
-		return buff.ptr(), nil
-	}
+// ipToRawSockaddrAny is a helper function that converts an assumed valid net.IP (IPv4) to *syscall.RawSockaddrAny.
+func ipToRawSockaddrAny(ip net.IP) *syscall.RawSockaddrAny {
+	testdetection.MustBeTesting()
 
-	// We tried 10 times and the buffer is still too small: give up.
-	return nil, errors.New("iteration limit reached")
-}
+	ip4 := ip.To4()
+	if ip4 == nil {
+		return nil
+	}
 
-// Constants for byte size conversion.
-const kilobyte uint32 = 1024
+	sa := new(syscall.RawSockaddrInet4)
+	sa.Family = syscall.AF_INET
+	copy(sa.Addr[:], ip4) // ip4 is already a 4-byte slice
 
-// buffer is a type that allows resizing a slice of any type to a given number of bytes.
-type buffer[T any] struct {
-	data []T
+	//nolint:gosec // Unsafe is required to manipulate pointers at the Win32 API level, only used in tests.
+	return (*syscall.RawSockaddrAny)(unsafe.Pointer(sa))
 }
 
-// byteCount returns the number of bytes in the buffer.
-func (b buffer[T]) byteCount() uint32 {
-	var t T
-	sizeOf := uint32(reflect.TypeOf(t).Size())
-	n := uint32(len(b.data))
-	return n * sizeOf
-}
+// fillFromTemplate fills the ipAdapterAddresses struct with the values from the mockIPAddrsTemplate struct for testing purposes.
+func (a *ipAdapterAddresses) fillFromTemplate(template *mockIPAddrsTemplate, next *ipAdapterAddresses) {
```

Comment:
It seems we only fill the public methods from that struct. Do you think this could be something that is in the `daemontests` subpackage too that we inject only when needed? (and so, it doesn’t end up in production code too?) That would cascade to `ipToRawSockaddrAny` too.